butterfree/load/writers/delta_writer.py

```diff
@@ -0,0 +1,164 @@
+from delta.tables import DeltaTable
+from pyspark.sql.dataframe import DataFrame
+
+from butterfree.clients import SparkClient
+from butterfree.configs.logger import __logger
+
+logger = __logger("delta_writer", True)
+
+
+class DeltaWriter:
+    """Control operations on Delta Tables.
+
+    Resposible for merging and optimizing.
+    """
+
+    @staticmethod
+    def _get_full_table_name(table, database):
+        if database:
+            return "{}.{}".format(database, table)
+        else:
+            return table
+
+    @staticmethod
+    def _convert_to_delta(client: SparkClient, table: str):
+        logger.info(f"Converting {table} to Delta...")
+        client.conn.sql(f"CONVERT TO DELTA {table}")
+        logger.info("Conversion done.")
+
+    @staticmethod
+    def merge(
+        client: SparkClient,
+        database: str,
+        table: str,
+        path: str,
+        merge_on: list,
+        source_df: DataFrame,
+        when_not_matched_insert_condition: str = None,
+        when_matched_update_condition: str = None,
+        when_matched_delete_condition: str = None,
+    ):
+        """
+        Merge a source dataframe to a Delta table.
+
+        By default, it will update when matched, and insert when
+        not matched (simple upsert).
+
+        You can change this behavior by setting:
+        - when_not_matched_insert_condition: it will only insert
+            when this specified condition is true
+        - when_matched_update_condition: it will only update when this
+            specified condition is true. You can refer to the columns
+        in the source dataframe as source.<column_name>, and the columns
+            in the target table as target.<column_name>.
+        - when_matched_delete_condition: it will add an operation to delete,
+            but only if this condition is true. Again, source and
+            target dataframe columns can be referred to respectively as
+            source.<column_name> and target.<column_name>
+        """
+        try:
+            full_table_name = DeltaWriter._get_full_table_name(table, database)
+
+            table_exists = client.conn.catalog.tableExists(full_table_name)
+            # table_is_delta = DeltaTable.isDeltaTable(client.conn, path)
+
+            if table_exists:
+                pd_df = client.conn.sql(
+                    f"DESCRIBE TABLE EXTENDED {full_table_name}"
+                ).toPandas()
+                provider = (
+                    pd_df.reset_index()
+                    .groupby(["col_name"])["data_type"]
+                    .aggregate("first")
+                    .Provider
+                )
+                table_is_delta = provider.lower() == "delta"
+
+                if not table_is_delta:
+                    DeltaWriter()._convert_to_delta(client, full_table_name)
```

Comment:
If I understood correctly, if the table exists, we're converting it to delta, right?

Since we're using Pandas, is there any change that the Pandas representation might consume too much memory based on the size of the table or this Pandas representation is just storing some metadata that doesn't get bigger when the table gets bigger?

Reply:
Yes, we are converting. Describe table is a very simple query, it's the same resource consumption no matter the size of the table. That is a trick found to correctly identify the table type without using path.